File: src/days/day8.rs
use std::collections::HashSet;
use std::fs;


pub fn solve() {
    println!("Day 8 Part 1: {}", part1());
    println!("Day 8 Part 2: {}", part2());
}

fn part1() -> u32 {
    let contents = fs::read_to_string("data/day8.txt").expect("File could not be read");
    let lines: Vec<&str> = contents.lines().collect();

    let mut next_instruction: i32 = 0;
    let mut visited: HashSet<i32> = HashSet::new();

    let mut accumulator: i32 = 0;

    while !visited.contains(&next_instruction) {
        visited.insert(next_instruction);
        let args: Vec<&str> = lines[next_instruction as usize].split_whitespace().collect();
        match args[0] {
            "acc" => {
                accumulator += args[1].parse::<i32>().unwrap();
                next_instruction += 1;
            },
            "jmp" => {
                next_instruction += args[1].parse::<i32>().unwrap();
            },
            "nop" => {
                next_instruction += 1;
            },
            _ => panic!("Found illegal instruction {0}", lines[next_instruction as usize]),
        }
    }

    accumulator as u32
}

fn part2() -> u32 {
    0
}

File: 2020/src/days/day8.py
def main():
    with open("data/day8.txt") as f:
        contents = f.read().strip()

    instructions = [line.split(" ") for line in contents.split("\n")]
    instructions = [(command, int(value)) for command, value in instructions]

    for i in range(len(instructions)):
        command, value = instructions[i]
        if command == "acc":
            continue

        print("Checking instruction %d..." % i, end="")
        if command == "jmp":
            halt_value = halts(instructions[:i] + [("nop", value)] + instructions[i + 1:])
            if halt_value:
                print("halts with value %d" % halt_value)
                return halt_value

        if command == "nop":
            halt_value = halts(instructions[:i] + [("jmp", value)] + instructions[i + 1:])
            if halt_value:
                print("halts with value %d" % halt_value)
                return halt_value

        print("doesn't halt")


def halts(instructions):
    visited = set()
    next_ptr = 0
    accumulator = 0

    while next_ptr not in visited:
        visited.add(next_ptr)
        if next_ptr == len(instructions):
            return accumulator
        command, value = instructions[next_ptr]

        if command == "acc":
            accumulator += value
            next_ptr += 1
        elif command == "nop":
            next_ptr += 1
        elif command == "jmp":
            next_ptr += value
        else:
            raise ValueError("Found unexpected command: %s", command)

    return False

if __name__ == "__main__":
    main()

